### PR TITLE
CHAT-782 Performance regression with chat web socket

### PR DESCRIPTION
**Pull request by boubaker**
This PR will add a new module for common classes in Server and client side.
This will give the possibility to define classes in Shared Classloader of Chat Client Tomcat that will integrate eXo Kernel Services an that will reuse the common classes in chat.

The performance regression is caused by this [statement](https://github.com/exo-addons/chat-application/pull/175/files#diff-b6237169f817d5151a0e674980530f0aL237) and [this one](https://github.com/exo-addons/chat-application/pull/175/files#diff-13a37d25665b223daf6ef8bd22e46b69L175) that ensures that user spaces are updated each time he displays a page. The fix is made to add user spaces only once in login time and when the user joins a space using a [Space listener](https://github.com/exo-addons/chat-application/pull/175/files#diff-6dfe34cb61d155dcc5acc4e7bb4e1515R1). This space listener uses Some Chat classes that aren't defined in Tomcat shared library, thus it can't be used by the Space Listener that is injected using eXo Kernel dependencies. So the new common JAR ensures that utils classes are in shared Tomcat library to be able to use it in Kernel components and plugins.

### CODE REVIEW

**Comment by trongtt** [comment]
@boubaker could you please give me short insight of the root problem and the idea to fix this ? If I understand well, is it about the cost of opening connection between chat client and server ?

**Comment by boubaker** [comment]
> @boubaker could you please give me short insight of the root problem and the idea to fix this ? If I understand well, is it about the cost of opening connection between chat client and server ?

The description is modified to add more details.

**Comment by boubaker** [comment]
> @boubaker what is the status of performances tests with this fix ?

It improves the performances, but we still have some problems, I'll let TQA give results on JIRA as usual.

**Comment by thomasdelhomenie** [comment]
ok thanks. So I guess we have to wait for additional updates and tests on this PR before validating it definitively ?

**Comment by boubaker** [comment]
> ok thanks. So I guess we have to wait for additional updates and tests on this PR before validating it definitively ?

If you have some remarks, don't hesitate

File: application/pom.xml [hunk 1]
```diff
@@ -24,12 +20,18 @@
...
       <groupId>org.exoplatform.addons.chat</groupId>
-      <artifactId>chat-server-embedded</artifactId>
-      <classifier>classes</classifier>
+      <artifactId>chat-common</artifactId>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.exoplatform.addons.chat</groupId>
+      <artifactId>chat-services</artifactId>
+      <type>jar</type>
```
Comment (thomasdelhomenie) [inline application/pom.xml]: Is this mandatory ?

File: services/src/main/java/org/exoplatform/addons/chat/listener/ServerBootstrap.java [hunk 2]
```diff
@@ -17,42 +17,66 @@
...
-  }
-
   private static final Log LOG = ExoLogger.getLogger(ServerBootstrap.class.getName());
 
+  private static String serverUrl;
+
+  /**
+   * Get mongo database name for current tenant if on cloud environment
+   */
+  public static String getDBName() {
```
Comment (thomasdelhomenie) [inline services/src/main/java/org/exoplatform/addons/chat/listener/ServerBootstrap.java]: Why not moving this class in ChatUtils of common module ? It is used in multiple locations, not only related to server bootstrap.
Same for saveSpaces method.

File: services/pom.xml [hunk 3]
```diff
@@ -66,8 +66,18 @@
...
-      <groupId>org.apache.commons</groupId>
-      <artifactId>commons-lang3</artifactId>
+      <groupId>org.exoplatform.addons.chat</groupId>
+      <artifactId>chat-common</artifactId>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.exoplatform.social</groupId>
+      <artifactId>social-component-core</artifactId>
+      <scope>compile</scope>
```
Comment (thomasdelhomenie) [inline services/pom.xml]: Shouldn't it be 'provided' ?

File: services/src/main/java/org/exoplatform/addons/chat/listener/UpdateUserStatusListener.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import org.exoplatform.services.user.UserStateModel;
+import org.exoplatform.services.user.UserStateService;
+
+import java.util.Calendar;
+
+
+/**
+ * This listener will be used to update chat status of current user in the case session destroyed.
+ */
+public class UpdateUserStatusListener extends Listener<ConversationRegistry, ConversationState> {
```
Comment (thomasdelhomenie) [inline services/src/main/java/org/exoplatform/addons/chat/listener/UpdateUserStatusListener.java]: Is this new listener related to the current issue ?

File: common/pom.xml [hunk 5]
```diff
@@ -0,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (thomasdelhomenie) [inline common/pom.xml]: Just wondering, shouldn't we move classes to services module instead of creating a new module ? It would avoid to have yet another module/artifact. WDYT ?

Comment (boubaker) [reply common/pom.xml]: In fact the chat-services.jar is not included in chatServer.war and thus we need a JAR module for both client and server.
We have three JAR modules:
* Chat common
* Chat client: chat-services
* Chat-server: chat-server-embedded (The name of the artifact is not good IMO)

File: services/src/main/java/org/exoplatform/addons/chat/listener/ServerBootstrap.java (hunk 2 above)
Comment (boubaker) [reply services/src/main/java/org/exoplatform/addons/chat/listener/ServerBootstrap.java]: The social artifacts aren't included in Server standalone while chat-common jar is included in Server standalone. So we don't have to add extra dependencies on Chat Server standalone just for better looking for code IMO.

File: application/pom.xml (hunk 1 above)
Comment (boubaker) [reply application/pom.xml]: Yes, instead of integrating chat-server artifact, we include only client JAR.
Or do you mean "<type>jar</type>" ? If yes, of course, it's not mandatory since it's the default type of artifact.

File: services/pom.xml (hunk 3 above)
Comment (boubaker) [reply services/pom.xml]: done

File: services/src/main/java/org/exoplatform/addons/chat/listener/UpdateUserStatusListener.java (hunk 4 above)
Comment (boubaker) [reply services/src/main/java/org/exoplatform/addons/chat/listener/UpdateUserStatusListener.java]: Nope, that's the problem in fact, this listener and its configuration was added in a JAR not included in Shared Tomcat lib and apparently it was ignored while changing the packaging as I told you in Chat. In fact it's this class:
https://github.com/exo-addons/chat-application/blob/fix/CHAT-782/exo-addons-chat-extension-config/src/main/java/org/exoplatform/addons/chat/api/UpdateUserStatusListener.java

**Comment by boubaker** [comment]
> I did some tests, it works fine in embedded mode (with and without docker), but not in standalone mode. I used the docker-compose file (just had to update docker files and catalog to change 1.6 to 1.7). It seems the application cannot reach the chat server.

You'll have to add an apache frontend that serves requests on both eXo and Chat Server to make it work in standalone mode.

File: common/pom.xml (hunk 5 above)
Comment (thomasdelhomenie) [reply common/pom.xml]: Yes, my point (not very clear I agree :)) was actually to have only one jar (chat-services) and use it everywhere (so to embed it also in chatServer.war). Does it make sense ?

File: services/src/main/java/org/exoplatform/addons/chat/listener/UpdateUserStatusListener.java (hunk 4 above)
Comment (thomasdelhomenie) [reply services/src/main/java/org/exoplatform/addons/chat/listener/UpdateUserStatusListener.java]: ok

File: application/pom.xml (hunk 1 above)
Comment (thomasdelhomenie) [reply application/pom.xml]: Yes I meant only "jar". We generally don't add the type for "jar" since it is the default value.

**Comment by thomasdelhomenie** [comment]
Works fine now in standalone mode :+1: but we have this warning during startup:

exo_1         | 2018-01-04 09:32:17,290 | WARN  | Can not update firstName/lastName to chatServer [o.e.a.c.listener.UpdateUserEventListener<Catalina-startStop-1>]

**Comment by boubaker** [comment]
> exo_1 | 2018-01-04 09:32:17,290 | WARN | Can not update firstName/lastName to chatServer [o.e.a.c.listener.UpdateUserEventListener]

this comes from the old disabled listener, fixed.
